org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/MatchLocator.java

```diff
@@ -3457,4 +3444,69 @@ protected boolean typeInHierarchy(ReferenceBinding binding) {
 	}
 	return false;
 }
+/*
+ * Go through the nest hierarchy of a type, to try to create a method handle for a method declaration.
+ * Does not try to create a method handle for the specified type, only for types higher in the nest hierarchy.
+ *
+ * See:
+ * https://github.com/eclipse-jdt/eclipse.jdt.core/issues/375
+ * https://bugs.eclipse.org/bugs/show_bug.cgi?id=468127
+ */
+private IJavaElement createHandleFromNestAncestors(AbstractMethodDeclaration method, IJavaElement type) throws JavaModelException {
+	BinaryType binaryType = getBinaryType(type);
+	if (binaryType != null && binaryType.isAnonymous()) {
+		boolean isJarOrClass = isJarOrClass(binaryType);
+		if (isJarOrClass) {
+			return findMethodInNestAncestors(method, binaryType);
+		} else {
+			// TODO: JAVA 9 - JIMAGE to be included later - currently assuming that only .class
+			// files will be dealt here.
+		}
+	}
+	return null;
+}
+
+/*
+ * Tries to find a method in the nest ancestors of a type. E.g. for Type$2$1, nest ancestors are: Type$1, Type
+ */
+private IJavaElement findMethodInNestAncestors(AbstractMethodDeclaration method, BinaryType binaryType) {
+	IJavaElement methodHandle = null;
+	int index = 0;
+	// loop until we find the method or arrive at the nested classes host, e.g. for Type$2$1 the "visited" types are: Type$1, Type
+	while (methodHandle == null && index != -1) {
+		String typeQualifiedName = binaryType.getTypeQualifiedName();
+		index = typeQualifiedName.lastIndexOf('$');
+		if (index != -1) {
+			typeQualifiedName = typeQualifiedName.substring(0, index);
+		}
```

Comment:
```suggestion
		} else {
			break;
		}
```

Reply:
I don't know if this works in all cases. It means we don't go to the main type and only visit nested types. If we go with this we'd have to adjust the comments.

I can't run into the hang when running the test locally.

Reply:
Me either, so I wonder if the code you've added runs into the endless loop here:
https://github.com/eclipse-jdt/eclipse.jdt.core/blob/6cc3c6a76d283e8d64a28f8fe7113159dd592e27/org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/MatchLocator.java#L556

Reply:
OK I see it, the loop is broken. It doesn't store the qualified type name that it wants to iterate over.